LFS-241: Add support for questionnaire sections

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-241)
**Testing branch**: `LFS-241-test`. Use the Demographics form there for an example of a deeply nested section and a section link.

This branch allows you to define "sections", where questions can be grouped (in later PRs, they will be able to be repeated & conditionally inserted).

New NodeTypes: 
- `lfs:SectionLink`, which defines a link to a section (useful if you wish to repeat a set of questions from a different questionnaire)
- `lfs:AnswerSection`, which groups together answers (useful later, when we need to repeat a section n times and want to keep the answers in a sane order).

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Add Sections functionality

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Move form entry rendering code into its own file

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Add support for section titles and descriptions.
Fix docs.

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Fix styling with help from @marta-

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Add saving/loading support.

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Fix support for SectionLinks

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Add missing newline

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Codestyle improvements

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Docfix

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -0,0 +1,82 @@
+  const headerVariant = (depth > 3 ? "body1" : ("h" + (depth+3)));
+
+  const titleEl = sectionDefinition["label"] && <Typography variant={headerVariant}>{sectionDefinition["label"]} </Typography>;
+  const descEl = sectionDefinition["description"] && <Typography variant="overline">{sectionDefinition["description"]} </Typography>
```
```suggestion
  const descEl = sectionDefinition["description"] && <Typography variant="caption">{sectionDefinition["description"]} </Typography>
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -0,0 +1,82 @@
+      <input type="hidden" name={`${sectionPath}/question`} value={sectionDefinition['jcr:uuid']}></input>
+      <input type="hidden" name={`${sectionPath}/question@TypeHint`} value="Reference"></input>
+
+      <Grid container direction="column" spacing={4} alignItems="stretch" justify="space-between" wrap="nowrap">
```
@OrdiNeu Should we define these attributes somewhere higher up in the hierarchy (in Form maybe) and propagate them to the Section, to ensure consistency and avoid code duplication?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -0,0 +1,82 @@
+  const { classes, depth, existingAnswer, path, sectionDefinition } = props;
+  const [sectionID] = useState((existingAnswer && existingAnswer[0]) || uuidv4());
+  const sectionPath = path + "/" + sectionID;
+  const headerVariant = (depth > 3 ? "body1" : ("h" + (depth+3)));
```
The reasoning behind this could be made easier to understand, something along the lines of:
```suggestion
  const headerVariant = (depth > (MAX_HEADING_LEVEL - MIN_HEADING_LEVEL) ? "body1" : ("h" + (depth+MIN_HEADING_LEVEL)));
```
where MAX_HEADING_LEVEL is 6 and MIN_HEADING_LEVEL comes from the form and is 1 + The heading level of the title.

Or at least add a comment :)

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Fix the passing around of props
Alter SectionLinks so they are specified via paths, which allows them to
be given during content loading via the content JSON

OrdiNeu committed:
LFS-241: Add support for questionnaire sections

Change Typography variant used in section descriptions by suggestion from @marta-

Co-Authored-By: Marta Girdea <marta.girdea@gmail.com>

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -0,0 +1,96 @@
+  alignItems: "stretch",
+  justify: "space-between",
+  wrap: "nowrap"
+};
```
Can this be moved to `QuestionnaireStyle.jsx`?

marta- replied:
Then the props can also be used in [Subject.jsx](https://github.com/ccmbioinfo/lfs/blob/846a64e6bef363696551900417b1c3592afe3b93/modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx#L123).

OrdiNeu replied:
That's a pretty good idea!

marta- replied:
> Then the props can also be used in [Subject.jsx](https://github.com/ccmbioinfo/lfs/blob/846a64e6bef363696551900417b1c3592afe3b93/modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx#L123).

After some discussion with @OrdiNeu it was decided not to reuse the props in Subject.jsx because semantically it's a diferent type of container. Additionally, Subject will no longer import QuestionnaireStyle.

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Move CONTAINER_PROPS to QuestionnaireStyle.jsx

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Rename CONTAINER_PROPS to FORM_ENTRY_CONTAINER_PROPS
Fix styles.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -193,8 +168,8 @@ function Form (props) {
-            .filter(([key, value]) => value['jcr:primaryType'] == 'lfs:Question')
-            .map(([key, questionDefinition]) => <Grid item key={key}>{displayQuestion(questionDefinition, key)}</Grid>)
+            .filter(([key, value]) => ENTRY_TYPES.includes(value['jcr:primaryType']))
+            .map(([key, questionDefinition]) => FormEntry(questionDefinition, ".", 0, data, key))
```
```suggestion
            .map(([key, entryDefinition]) => FormEntry(entryDefinition, ".", 0, data, key))
```

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Improve error handling for invalid sections.
Code cleanup.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormEntry.jsx:
```diff
@@ -0,0 +1,115 @@
+  } else if (SECTION_TYPES.includes(entryDefinition["jcr:primaryType"])) {
+      return displaySection(entryDefinition, path, depth, existingAnswers, key);
+  }
+}
```
FTR, [LFS-274](https://phenotips.atlassian.net/browse/LFS-274) was created for refactoring this code.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -0,0 +1,86 @@
+const MIN_HEADING_LEVEL = 3;
+
+/**
+ * Component that consists of a few sections, and optionally has some criteria to its display
```
Doesn't seem right...

```suggestion
 * Component responsible for displaying a (sub)section of a questionnaire, consisting of a title, a description,
 * a list of questions and/or subsections, and may have some conditions to its display.
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -0,0 +1,86 @@
+
+  const titleEl = sectionDefinition["label"] && <Typography variant={headerVariant}>{sectionDefinition["label"]} </Typography>;
+  const descEl = sectionDefinition["description"] && <Typography variant="caption" color="textSecondary">{sectionDefinition["description"]} </Typography>
+  const hasPadding = titleEl || descEl;
```
```suggestion
  const hasHeader = titleEl || descEl;
```

OrdiNeu replied:
I disagree -- this variable also affects the amount of space between this section and the first question immediately after it (via the `classes.paddedSection` style). We wanted this to be disable-able since we wanted the concept of a section that exists but blends in seamlessly with the rest of the form.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -0,0 +1,86 @@
+  return (
+    <Grid item className={hasPadding && classes.paddedSection}>
+      <input type="hidden" name={`${sectionPath}/jcr:primaryType`} value={"lfs:AnswerSection"}></input>
+      <input type="hidden" name={`${sectionPath}/question`} value={sectionDefinition['jcr:uuid']}></input>
```
`question` is not declared in the nodetype.

Should it be `question` or `section`?

sdumitriu replied:
Actually it is, since it inherits from `lfs:Answer`.

But should it inherit from `lfs:Answer`? It is not a kind of answer, and I don't think it has a value. It is just a logical grouping of answers, corresponding to a section, so I would instead name it as `SectionAnswers`, not inheriting `Answer`, and having a `section` property instead of a `question` property.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -0,0 +1,86 @@
+  const hasPadding = titleEl || descEl;
+
+  return (
+    <Grid item className={hasPadding && classes.paddedSection}>
```
```suggestion
    <Grid item className={hasHeader && classes.labeledSection}>
```

sdumitriu reviewed (changes requested):
I think referenced sections are not as easy to implement and should be moved out of this PR.

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Apply suggestions by @sdumitriu
Move lfs:AnswerSection to its own node type, independent of lfs:Answer
Docfix on Section
Exit early for invalid sections before trying to find a corresponding
answer

OrdiNeu committed:
LFS-241: Add support for questionnaire sections
Bugfix on the wrong prop type for existingAnswer and wrong parse for
Sections
Remove SectionLinks out of this PR as suggested by @sdumitriu

marta- reviewed (approved):
Looks good.

sdumitriu reviewed (approved):
One thing that is broken is that the quicksearch doesn't match text entered in sections. The index will have to be adjusted.
